packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+            e.setHeader("Package files must be typed: strict");
+        }
+    }
+
+    {
+        UnorderedMap<core::NameRef, core::LocOffsets> importedNames;
+        for (auto imported : package.importedPackageNames) {
+            auto it = packageInfoByMangledName.find(imported.mangledName);
+            if (it == packageInfoByMangledName.end()) {
+                if (auto e = ctx.beginError(imported.loc, core::errors::Packager::PackageNotFound)) {
+                    e.setHeader("Cannot find package `{}`", absl::StrJoin(imported.fullName, "::", NameFormatter(ctx)));
+                }
+                continue;
+            }
+
+            if (importedNames.contains(imported.mangledName)) {
+                if (auto e = ctx.beginError(imported.loc, core::errors::Packager::InvalidImportOrExport)) {
+                    e.setHeader("Duplicate package import `{}`",
+                                absl::StrJoin(imported.fullName, "::", NameFormatter(ctx)));
+                    e.addErrorLine(core::Loc(ctx.file, importedNames[imported.mangledName]),
+                                   "Previous package import found here");
+                }
+            } else {
+                importedNames[imported.mangledName] = imported.loc;
+                const auto &found = *it->second;
+
+                ast::ClassDef::RHS_store exportedItemsCopy;
+                for (const auto &exported : found.exportedItems) {
+                    exportedItemsCopy.emplace_back(exported.deepCopy());
+                }
+
+                ENFORCE(!imported.fullName.empty());
+                // Create a module for the imported package that sets up constant references to exported items.
+                // Use proper loc information on the module name so that `import Foo` displays in the results of LSP
+                // Find All References on `Foo`.
+                importedPackages.emplace_back(ast::MK::Module(imported.loc, core::Loc(ctx.file, imported.loc),
+                                                              imported.getFullNameLiteral(imported.loc), {},
+                                                              std::move(exportedItemsCopy)));
+            }
+        }
+    }
+
+    // Include Kernel in the package.
+    importedPackages.emplace_back(ast::MK::Send1(core::LocOffsets::none(), ast::MK::Self(core::LocOffsets::none()),
+                                                 core::Names::include(), name2Expr(core::Names::Constants::Kernel())));
+
+    auto packageNamespace =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(package.name.mangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(importedPackages));
+
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    rootKlass->rhs.emplace_back(move(packageNamespace));
```

Comment:
```suggestion
    auto rootKlass = ast::cast_tree_nonnull<ast::ClassDef>(file.tree);
    rootKlass.rhs.emplace_back(move(packageNamespace));
```

Reply:
There are a handful of places where you could replace `cast_tree` followed by `ENFORCE` with `cast_tree_nonnull`. Feel free to change them too, but it's not blocking. I just think it's a little more concise.

Reply:
Agreed. I had to update this code several time as Trevor landed TreePtr changes. Thanks for the pointers!